code/game/objects/items/devices/PDA/PDA.dm

```diff
@@ -903,6 +903,29 @@ GLOBAL_LIST_EMPTY(PDAs)
 		playsound(src, 'sound/machines/pda_button1.ogg', 50, TRUE)
 	return TRUE
 
+/obj/item/pda/pre_attack(obj/target, mob/living/user, params)
+	if(!ismachinery(target))
+		return ..()
+	var/obj/machinery/target_machine = target
+	if(!target_machine.panel_open && !istype(target, /obj/machinery/computer))
+		return ..()
+	if(!istype(cartridge, /obj/item/cartridge/virus/clown))
+		return ..()
+	var/obj/item/cartridge/virus/installed_cartridge = cartridge
+
+	if(installed_cartridge.charges <=0)
+		to_chat(user, "<span class='notice'>Out of charges.</span>")
+		return ..()
+	to_chat(user, "<span class='notice'>You upload the virus to the airlock controller!</span>")
```

Comment:
Is this message going to be displayed as-is even when hacking things other than airlocks?

Reply:
For now it is limited to cartridge/virus/clown so only airlocks will display it

Reply:
Not the case in the current state of the PR as far as I can tell.

![image](https://user-images.githubusercontent.com/6917698/123986718-d974e200-d9c6-11eb-8345-46805e75a5f5.png)

Reply:
That's what I get for copypasting tg lmao. Will fix right away.

Reply:
Alright with next commit this shouldn't issue any longer.